Approved/remove-new-array.md

```diff
@@ -0,0 +1,84 @@
+---
+title: Remove default-initialized new array expressions
+description: The `new Type[N]` array expression is deprecated and removed from Q#.
+author: Sarah Marshall
+date: October 28, 2020
+---
+
+# Proposal
+
+The `new Type[N]` expression, which creates an array of `N` values using the default value of `Type`, is deprecated and removed from Q#.
+
+# Justification
+
+The existence of `new Type[N]` has negative effects on the language.
+
+The most important negative effect is the assumption that every type has a default value.
+This is not a reasonable assumption, because it is not possible to define useful default values of types like `Qubit` and `a -> b`.
+Their current default values are invalid and will trigger an error if they are used.
+But the syntax is the same both for types with valid default values, and those without, which is misleading: while `new Int[3]` is a well-defined value whose items can be used safely, `new Qubit[3]` creates a potential trap that will crash the program if any item in the array is used before being set to a valid value first.
+
+This assumption is also invalid in the presence of uninhabited types such as `Void`.
+If Q# requires that every type has a default value, then uninhabited types are not possible to express properly.
+This can even cause subtle bugs in the soundness of a type system, as demonstrated by [this bug in Java's generics](https://hackernoon.com/java-is-unsound-28c84cb2b3f) that happened because of the existence of a value (specifically `null`) for a type that should have been uninhabited.
+
+Finally, the use of the `new` keyword is inconsistent with other ways to create values in Q#.
+`new` is required only when creating a default-initialized array, not when using a square bracket array literal or when creating values of other types.
+
+# Description
+
+## Current Status
+
+Q# currently supports this syntax:
+
+```qsharp
+// Create an empty array of type Int[].
+let empty = new Int[0];
+
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = new Bool[10];
+```
+
+## Proposed Modification
+
+The syntax above will be removed.
+The same functionality can be accomplished with functions in the standard library:
+
+```qsharp
+// Create an empty arra of type Int[].
+let empty = EmptyArray<Int>();
+
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = ConstantArray(10, false);
+```
+
+# Implementation
+
+The library functions `EmptyArray` and `ConstantArray` are already implemented.
```

Comment:
Does this mean that we'll need to import a library every time we create an array?

Reply:
You can still create array literals like `[1, 2, 3]` without library functions. To create empty arrays or arrays consisting only of a repeated constant, you need to use `EmptyArray` or `ConstantArray` from the `Microsoft.Quantum.Arrays` namespace. We could consider moving them into `Microsoft.Quantum.Core` so that it's not necessary to `open Microsoft.Quantum.Arrays` to access them.

Reply:
@cgranade, do you have any thoughts on adding `EmptyArray` and/or `ConstantArray` to `Microsoft.Quantum.Core`?

Reply:
To clarify, with the current state of affairs, no library needs to be imported to use `EmptyArray` and `ConstantArray`, since both are part of the Q# standard library and hence provided by the packages made available  by default in new projects (whether Q# standalone command-line, Q# standalone notebooks, Q# + Python, or Q# + .NET). That said, one does need an `open` statement at the moment to make the `Microsoft.Quantum.Arrays` _namespace_ available; our language server already suggests doing so if a user has forgotten in VS Code or VS 2019.

With respect to moving `EmptyArray` and `ConstantArray` to the Microsoft.Quantum.Core namespace, I'm somewhat leery since everything in Microsoft.Quantum.Core is automatically opened and thus available to all Q# programs even without an `open` statement. That's a significant exception to the use of namespaces to document where things come from, organize code, and so forth — that exception is really helpful in cases like `Length`, but in general I think we should make as few exceptions of that form as is reasonable. Thus, while it may make sense to move those two functions, I'd suggest that before doing so, we expand the API design principles to be more clear about the role of the Microsoft.Quantum.Core namespace in particular.

Reply:
My ideal solution, which doesn't fit Q#'s naming convention right now, is to auto-open `Microsoft.Quantum.Arrays` using an alias - so every Q# namespace has an implicit

```qsharp
open Microsoft.Quantum.Arrays as Array;
```

Then I would rename `EmptyArray` to `Empty`, and `ConstantArray` to `Constant`, and we could do `Array.Empty<T>()` and `Array.Constant(n, value)`.

Reply:
That feels eerily close to modules...

Reply:
Yes, this is one of several reasons why I think Q# should replace namespaces with modules. Having access to all functions related to arrays, lists, etc. in F#, without polluting the global namespace, is something I miss in Q#.

Reply:
I think having to construct a constant array is common enough that the namespace would have to be opened almost everywhere and we should find a better solution. Related, also IndexRange is one thing that we might consider having available without having the explicitly open a namespace. We could certainly mitigate by having the M.Q.Array namespace opened in our project templates like M.Q.Canon. But even then, having an increasingly long list of namespaces at the top of each file even for a small example to me is detrimental to the impression that everything is simple. I hence think we should have more auto-opening than we currently do. 
I am curious about the arguments against having namespaces automatically open under an alias by default, with the option to disable that with a project configuration. @cgranade @guenp @msoeken, thoughts?

Reply:
Thanks for pinging me, @bettinaheim! Following our offline discussion, if I could try to summarize, it would be good to have a way to declare aliases as being automatically opened with new projects and notebooks before moving forward with verbose names like `Microsoft.Quantum.Arrays.EmptyArray`. For example, if `Array.Empty<Int>()` worked by default using auto-open functionality, and could be reduced further to `Empty<Int>()` with `open` statements, then that would make the library-based approach a better option. 

Two options for doing so that may be good to include in a new proposal: 

- Allow "re-exporting" namespace aliases declared with `open ... as` statements.
- Allow defining namespace aliases in csproj file properties.

Reply:
> and could be reduced further to `Empty<Int>()` with open statements, then that would make the library-based approach a better option.

In my opinion, `Empty<T>()` is confusing and should be avoided. If we shorten the names to make the aliases read better, I'd like to see Q# adopt something like F#'s `[<RequireQualifiedAccess>]` instead.

Reply:
Let's support empty array literals (`[]`) as part of implementing this proposal. Type inference based on the first usage is something we should do sooner or later in any case, so let's start with this.

Reply:
💕 for empty array literals!